### ISSUE TITLE
Add Top-K Motif Discovery Tutorial

### ISSUE THREAD

**Issue by seanlaw** [TDAmeritrade/stumpy#518]
Demonstrate the use of `stumpy.motifs` and `stumpy.match`

**Comment by seanlaw** [comment]
In the paper, it says:

> we tested our method on an electrical load measurement dataset [21]. The dataset consists of electrical load measurements for individual appliances (and an aggregated load) from households in United Kingdom. Five appliances are considered: fridge-freezer, freezer, tumble dryer, dishwasher,
and washing machine. The data was collected from April 19, 2014 to May 15, 2014, where the length is 17,000. The subsequence length was set to 4 hours. 

Cross referencing the cleaned version of the [REFIT dataset](https://pureportal.strath.ac.uk/en/datasets/refit-electrical-load-measurements), it was found that only "CLEAN_House3.csv" [in the "Access Dataset > REFIT Power Data - 111215" ](https://pureportal.strath.ac.uk/files/45646018/REFITPowerData111215.7z) contained the aforementioned appliances. It was not possible to reproduce the paper's data preprocessing steps but this should be sufficient:

```
%matplotlib inline

import matplotlib.pyplot as plt
import pandas as pd

plt.style.use('https://raw.githubusercontent.com/TDAmeritrade/stumpy/main/docs/stumpy.mplstyle')

start_date, end_date = '2014-04-19', '2014-05-15'
df = pd.read_csv("CLEAN_House3.csv")
df["Time"] = pd.to_datetime(df["Time"])
colnames = {
    "Appliance1": "Toaster",
    "Appliance2": "Fridge-Freezer",
    "Appliance3": "Freezer",
    "Appliance4": "Tumble Dryer",
    "Appliance5": "Dishwasher",
    "Appliance6": "Washing Machine",
    "Appliance7": "Television",
    "Appliance8": "Microwave", 
    "Appliance9": "Kettle",
}
df = (df.rename(colnames, axis="columns")
        .loc[:, ["Time",
                 "Fridge-Freezer",
                 "Freezer", 
                 "Tumble Dryer", 
                 "Dishwasher",
                 "Washing Machine"]]
        .query('Time >= @start_date and Time <= @end_date')
        .groupby(pd.Grouper(key='Time', freq="T")).sum()
     )

for colname in df.columns:
    plt.plot(df[colname])
```

**Comment by SaVoAMP** [comment]
Hey!

Since we were discussing this topic in detail [here](https://github.com/TDAmeritrade/stumpy/discussions/494) and have already generalized the multidimensional motif discovery to finding the top k motifs and top k nearest neighbors, I would like to help with this tutorial!

**Comment by seanlaw** [comment]
@SaVoAMP Sounds good! I will assign it to you

**Comment by SaVoAMP** [comment]
Should I just use the REFIT dataset for the tutorial?

If so, do you think it would suffice to just use the dataset locally and refer to the "download instructions" in a comment?

if not, since the download is a 489 MB archive, I can't import it "on the fly".

Sadly I don't have any servers to host the ´CLEAN_House3.csv´ file on except my own github account.
I presume that the .csv file won't be part of the repo? Maybe [Git LFS](https://git-lfs.github.com/) could be a solution here?

Do you have a link with the cleaned up version of the file that I could use for the tutorial, or should I create my own separate github project containing only the file? 
Or would GitHub be unsuitable for that purpose? 
I'm asking because my software engineering professor mentioned that GitHub/GitLab are not suited for filehosting and only sourcecode should be uploaded there. I don't know to what extent he ment that.
Or would you suggest using another dataset instead?

**Comment by SaVoAMP** [comment]
All right, thank you!

How do you want to visualize the motif pair? Should we plot it inside the (currently very colorful) graph or should I make a zoom-in of the motif pair and only plot that? 
As soon as we get matches we need to plot them inside the graph anyway. But if I would visualize all appliances in the same color so that we can see the motif pair and its nearest neighbors with another color, then we could not tell which appliance belonged to which time series. 
I would suggest to plot 5 subplots (one for each appliance) and afterwards visualize the motifs and matches with another color, as I have done here, for example:
 
![grafik](https://user-images.githubusercontent.com/95933072/156423179-28bcb323-1739-44ff-bbaf-1a7778166b67.png)
This was a 30-dimensional data set and after selecting the k dimensions where the motif is represented, I only plotted those dimensions with all its matches. The motif pair can be seen in red.

But the problem with our dataset might be that there are a plenty of spikes in each dimension which may look odd when plotted above each other 🤔

**Comment by seanlaw** [comment]
For things like this, It's difficult for me to judge without seeing the possible outcomes in context. If it's not too much work, is it possible to submit a PR with a notebook that displays/annotates the pros/cons of each approach to visualizing the results?

In case it matters, we don't necessarily need to use all 9 dimensions from the `CLEAN_House3.csv` dataset. We could always choose a subset where only 2 dimensions (i.e., washer + dryer) would be identified by MDL.

**Comment by SaVoAMP** [comment]
Our time series has the length 37,440. But the authors claim that their length is 17,000. If I understand everything correctly, you were converting that to minutes. 
I'm trying to plot the motif pair but stuggling since I only work with data frames since the beginning of my thesis. 
Since I was totally confused by the differing timesteps, I was only plotting "data points" on the x-axis instead of time (until now). That's why I am struggling to plot the time series in timesteps together with the motif now, since my indices aren't in time steps. Probably it will take some time until I've got it 😆 So far, I'm only managing to decently plot the time series that contain the motif. But as soon as I draw in the motif, everything breaks down 😄 I'll make a pull request as soon as I have a halfway decent plot!

**Comment by seanlaw** [comment]
> Our time series has the length 37,440. But the authors claim that their length is 17,000. If I understand everything correctly, you were converting that to minutes.

I think that's right. In `.groupby(pd.Grouper(key='Time', freq="T")).sum()`, I'm basically grouping data points to within each minute and then summing up all of the energy consumption during that minute in order to reduce the overall data size.

> I think we might have the problem, that the flat areas are more similar than the spikes that we want to catch.

I see. That's a nasty problem to have. I wonder if we can cheat/overcome this by adding ONE additional dimension that is simply the aggregate energy usage (I'm hoping that this will help reduce the flat areas)?

**Comment by SaVoAMP** [comment]
I don't think so since the similarity still holds in the other dimensions. I think that they will still be chosen and the additional one ignored.

**Comment by seanlaw** [comment]
What about adding white noise to constant regions?

**Comment by SaVoAMP** [comment]
Puhh, difficult to tell, but we could try. If it were a one-dimensional problem I would have tried it with annotation vectors. But I don't know if this is possible for the multidimensional case since each dimension needs to get annotated separately.

**Comment by seanlaw** [comment]
> Puhh, difficult to tell, but we could try. If it were a one-dimensional problem I would have tried it with annotation vectors. But I don't know if this is possible for the multidimensional case since each dimension needs to get annotated separately.

I don't know why I didn't think of this earlier but we could simply replace the constant regions of our time series with `np.nan`! This ensure that these regions MUST have a distance (in the distance profile) of `np.inf` and they would not be selected

**Comment by SaVoAMP** [comment]
yeah I could try, but I think that may eliminate too much. I think its better to set only zero values to `np.nan`. But I'm also not sure if this is too much. Depends on the length of the subsequence that we set 🤔

**Comment by SaVoAMP** [comment]
I'm not sure if we should eliminate regions that are only constant. I think it may be better to only set those regions to `np.nan` that are constantly zero since this means that there is no electrical power demand. With my other data set there were many problems when trying to eliminate some regions. I think we may eliminate too much since it looks like as if some appliances have the same electrical power demand over some time period and I'm not sure if we want to eliminate them. I think it's pretty difficult to find meaningful motifs here 🤔  I'll try experimenting a bit over the next few days.

What would you think of explaining the `mmotifs` function on a simplified example first and afterwards loading the difficult data set and trying to explore further, as you have done [here](https://stumpy.readthedocs.io/en/latest/Tutorial_Time_Series_Chains.html)? I could use the data that I used for the unit tests.

**Comment by seanlaw** [comment]
> What would you think of explaining the mmotifs function on a simplified example first and afterwards loading the difficult data set and trying to explore further, as you have done [here](https://stumpy.readthedocs.io/en/latest/Tutorial_Time_Series_Chains.html)? I could use the data that I used for the unit tests.

In general, I've avoided using simplified examples. However, as you've pointed out, there are times where it may be useful and I am not opposed to it as long as we follow it with a real-world example (a la REFIT data).

**Comment by SaVoAMP** [comment]
Somehow the partial sequence length looks very small. But the authors say that they used a subsequence length of 4 hours. But since you have changed the timestamp to minutes, I should use `m=4*60` right?

**Comment by seanlaw** [comment]
> Somehow the partial sequence length looks very small. But the authors say that they used a subsequence length of 4 hours. But since you have changed the timestamp to minutes, I should use `m=4*60` right?

That sounds about right to me.

**Comment by SaVoAMP** [comment]
I'm not sure what I'm doing wrong but it doesn't seem as if anything is eliminated.

**Comment by seanlaw** [comment]
@SaVoAMP At this stage, given that we don't know what dataset the authors had used and what preprocessing steps were taken, I don't think we can limit ourselves to reproducing their work anymore due to the lack of information. My thought process was more like "for the most part, humans tend to dry their laundry right after they wash it" so there's probably a motif there and maybe we can pick it up with matrix profiles. Maybe not? That was the extent of it.

**Comment by SaVoAMP** [comment]
Alright! 
My problem is that setting values with zero standard deviation to `np.nan` doesn't work and I can't explain why. I wrote two if-conditions and when I debug, I also get into the conditions. But when setting the values to `np.nan` it doesn't work, although the `constant_idx` vector is occupied. I probably need to take a closer look at the data frame operations here.

**Comment by seanlaw** [comment]
@SaVoAMP I think it would be better to continue this conversation over in your open PR so that there is continuity of the discussion. Let's continue over there if you don't mind.
